slides/advanced-02-model-tuning.qmd

```diff
@@ -0,0 +1,1064 @@
+#| code-line-numbers: "|1|6|"
+show_best(nnet_res, metric = "brier_class") |> 
+  relocate(.metric, mean) 
+```
+
+## Choose a parameter combination `r hexes(c("tune"))`
+
+Create your own tibble for final parameters or use one of the `tune::select_*()` functions:
+
+```{r}
+#| label: nnet-metrics-best
+#| code-line-numbers: "|1|6|8-10|14-17|"
+nnet_best <- select_best(nnet_res, metric = "brier_class")
+nnet_best
+
+collect_metrics(nnet_res) |> 
+  inner_join(nnet_best) |> 
+  select(.metric, mean)
+```
+
+## Checking (Approximate) Calibration `r hexes(c("tune", "probably"))`
+
+```{r}
+#| label: nnet-cal-plot
+#| output-location: column
+#| out-width: 90%
+#| fig-width: 5
+#| fig-height: 5
+
+nnet_holdout_pred <-
+  nnet_res |> 
+  collect_predictions(
+    parameters = nnet_best
+  )
+
+nnet_holdout_pred |>
+  cal_plot_windowed(
+    truth = class,
+    estimate = .pred_event,
+    window_size = 0.2,
+    step_size = 0.025,
+  )
+```
+
+
+::: notes
+This plot _pools_ the out-of-sample predictions. Our Brier estimate is the mean of 10 data sets.
+
+The curve is approximate for that reason. 
+:::
+
+
+## Multiple goals
+
+Optimizing the Brier score optimizes for accuracy and calibration of the class probabilities. 
+
+<br> 
+
+That's counter-productive to finding the rare class "event" events.
+
+<br>
+
+Can we find a way to optimize multiple metrics at once? 
+
+<br> 
+
+There are a few ways, and we'll focus on [desirability functions](https://aml4td.org/chapters/cls-metrics.html#sec-cls-multi-objectives). 
+
+## Desirability functions
+
+We create simple functions to translate our variable to [0, 1]; 1.0 is most desirable.
+
+
+```{r}
+#| label: three-desirability-examples
+#| echo: false
+#| out-width: 70%
+#| fig-width: 10
+#| fig-height: 3
+library(desirability2)
+
+p_roc <-
+  tibble(AUC = seq(0.3, 1.05, length.out = 1000)) |>
+  mutate(d_ROC = d_max(AUC, low = 1 / 2, high = 1)) |>
+  ggplot(aes(AUC, d_ROC)) +
+  geom_line() +
+  geom_vline(xintercept = 1 / 2, col = "red", lty = 2) +
+  geom_vline(xintercept = 1, col = "green", lty = 2) +
+  labs(x = "ROC AUC", y = "Desirability", title = "Goal: Maximize")
+
+p_brier <- 
+  tibble(brier = seq(-0.05, 0.30, length.out = 1000)) |>
+  mutate(d_brier = d_min(brier, low = 0, high = 0.25)) |>
+  ggplot(aes(brier, d_brier)) +
+  geom_line() +
+  geom_vline(xintercept = 1 / 4, col = "red", lty = 2) +
+  geom_vline(xintercept = 0, col = "green", lty = 2) +
+  labs(x = "Brier Score", y = NULL, title = "Goal: Minimize")
+
+p_spec <- 
+  tibble(spec = seq(-0.05, 1.05, length.out = 1000)) |>
+  mutate(d_spec = d_box(spec, low = 0.8, high = 1)) |>
+  ggplot(aes(spec, d_spec)) +
+  geom_line() +
+  labs(x = "Specificity", y = NULL, title = "Goal: Constrain")
+
+p_roc + p_brier + p_spec
+```
+
+We can combine them with a geometric mean. 
+
+
+## Multimetric optimization `r hexes(c("tune"))`
+
+
+```{r}
+#| label: nnet-metrics-desire-sens
+#| code-line-numbers: "|1|3|4|5|11|12-13|"
+show_best_desirability(
```

Comment:
These (and other) results are _very_ busy. I played a lot with using `gt()` or `kable()` tables but this seems to me to be the best option. 

However, feel free to suggest something else.